alphadia/workflow/optimizer.py

```diff
@@ -0,0 +1,147 @@
+        self.estimator = "mz"
+        self.df = "fragment"
+        self.ids = []
+
+    @check_if_optimum_found
+    def check_stopping_conditions(self):
+        """Optimization should stop if continued narrowing of the MS2 tolerance is not improving the number of precursor identifications.
+        This function checks if the previous rounds of optimization have led to a meaningful improvement in the number of identifications.
+        If so, it continues optimization and appends the proposed new tolerance to the list of tolerances. If not, it stops optimization and sets the optimal tolerance attribute.
+
+        Notes
+        -----
+            Because the check for an increase in identifications requires two previous rounds, the function will also initialize for another round of optimization if the round variable is less than 2.
+
+            Once the optimal tolerance is set, the optimizer will no longer check for improvements.
+
+        Returns
+        -------
+        stop_optimizaton: bool
+            True if there has been no recent improvement in the number of identifications and optimization should stop. False otherwise.
+
+        """
+
+        if (
+            self.round < 2
+            or self.ids[-1] > 1.1 * self.ids[-2]
+            or self.ids[-1] > 1.1 * self.ids[-3]
+        ):
+            self.reporter.log_string(
+                f"❌ MS2: optimization incomplete in round {self.round}. Will search with tolerance {self.proposed_new_tolerance}.",
+                verbosity="progress",
+            )
+            stop_optimizaton = False
+
+        else:
+            self.optimal_tolerance = self.tolerances[np.argmax(self.ids)]
+            self.end_of_optimization_message()
+            stop_optimizaton = True
+
+        return stop_optimizaton
+
+    @check_if_optimum_found
+    def initiate(self):
+        """If the optimization continues (which will use the proposed new tolerance), the round variable should be incremented and the proposed new tolerance should be added to the list of tolerances."""
+        self.tolerances.append(self.proposed_new_tolerance)
+        self.round += 1
+
+    @check_if_optimum_found
+    def update(self, precursors_df: pd.DataFrame, proposed_new_tolerance: float):
+        """This function appends the number of precursors identified to the list of identifications.
+        It also saves the proposed new tolerance, which will be added to the list of tolerances if the optimization continues after check is called.
+
+        Parameters
+        ----------
+
+        precursors_df: pd.DataFrame
+            The filtered dataframe of precursors obtained after FDR correction.
+
+        proposed_new_tolerance: float
+            The tolerance which will be used for the next round of optimization (if optimization continues).
+
+        """
+        self.ids.append(len(precursors_df))
+        self.proposed_new_tolerance = proposed_new_tolerance
+
+    def end_of_optimization_message(self):
```

Comment:
no need to separate this into a function now that we remove it from the decorator.

Reply:
Indeed.